### PR TITLE
Fix #380: Hi fi input interaction views

### LINKED ISSUE #380: Entering an exploration state sometimes scrolls past the content
If a state card has a text-based interaction, sometimes Android will scroll the user partly down past the content card (likely because the text input is focused). We shouldn't focus the interaction by default.

@rt4914 and I also noticed when playing through state cards that sometimes when selecting a text input to open the keyboard, the keyboard won't open on the first tap and the interaction will be scrolled off screen.

We need to generally fix jank in this area to ensure the experience is solid.

## Issue Work Items

- [ ] Whenever state is changed the scroll view should start from the top everytime.
- [ ] On click of input type interaction item, it requires two clicks to display keyboard, which should actually be just a single click.

### PR DESCRIPTION
**Pull request by nikitamarysolomanpvt**
## Explanation
Whenever state is changed the scroll view should start from the top everytime.For this removed the focus from edittext.

## Checklist
- [x] Whenever state is changed the scroll view should start from the top everytime.
- [x] On click of input type interaction item, it requires two clicks to display keyboard, which should actually be just a single click.
## Reference Mocks
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/4d4e4bd8-3755-41fc-bf83-3dcff5d5eee1/Lesson-1-a-ii-Exploration-Player-Text-Input-No-Ans
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/eeed0f6c-54ca-4371-ab23-5199abd87f28/Lesson-1-a-ii-Exploration-Player-Text-Input-Keyboa
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/d66097bf-8421-41d8-bfda-2450d6e98dea/Lesson-1-a-i-Exploration-Player-Fraction-Input-No-
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/249c3349-605b-4569-bbe9-1a017ab90101/Lesson-1-a-i-Exploration-Player-Fraction-Input-No-
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/c4b0dfb3-53b4-491e-a759-818e096183fd/Lesson-1-a-v-Exploration-Player-Numeric-Input-No-A

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Review by rt4914 — CHANGES_REQUESTED** [review]
Currently, there is one case in which the new state does not start from the top-position:

add this line
`binding.stateRecyclerView.smoothScrollToPosition(0)`

at the end of `StateFragmentPresenter` -> `processEphemeralStateResult` function.

**Comment by nikitamarysolomanpvt** [comment]
> Currently, there is one case in which the new state does not start from the top-position:
> 
> add this line
> `binding.stateRecyclerView.smoothScrollToPosition(0)`
> 
> at the end of `StateFragmentPresenter` -> `processEphemeralStateResult` function.

I fixed input view focus issue because of which the scroll was happening.
as per your suggestion added `binding.stateRecyclerView.smoothScrollToPosition(0)` in processEphemeralStateResult

File: app/src/main/java/org/oppia/app/customview/interaction/FractionInputInteractionView.kt [hunk 1]
```diff
@@ -26,11 +31,19 @@ class FractionInputInteractionView @JvmOverloads constructor(
...
+    showSoftKeyboard(v, context)
   } else {
     hint = hintText
-    setTypeface(typeface, Typeface.ITALIC)
+    if (text.isEmpty()) setTypeface(typeface, Typeface.ITALIC)
+    hideSoftKeyboard(v, context)
+  }
+
+  override fun onKeyPreIme(key_code: Int, event: KeyEvent): Boolean {
+    if (event.getKeyCode() == KEYCODE_BACK && event.getAction() == ACTION_UP)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/customview/interaction/FractionInputInteractionView.kt]: Nit: Update to `if (event.keyCode == KEYCODE_BACK && event.action == ACTION_UP)`

File: app/src/main/java/org/oppia/app/customview/interaction/NumericInputInteractionView.kt [hunk 2]
```diff
@@ -26,11 +31,20 @@ class NumericInputInteractionView @JvmOverloads constructor(
...
+    showSoftKeyboard(v, context)
   } else {
     hint = hintText
-    setTypeface(typeface, Typeface.ITALIC)
+    if (text.isEmpty()) setTypeface(typeface, Typeface.ITALIC)
+    hideSoftKeyboard(v, context)
+  }
+
+  override fun onKeyPreIme(key_code: Int, event: KeyEvent): Boolean {
+    if (event.getKeyCode() == KEYCODE_BACK && event.getAction() == ACTION_UP)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/customview/interaction/NumericInputInteractionView.kt]: Ditto as above

File: app/src/main/java/org/oppia/app/customview/interaction/TextInputInteractionView.kt [hunk 3]
```diff
@@ -25,11 +28,20 @@ class TextInputInteractionView @JvmOverloads constructor(
...
+    showSoftKeyboard(v, context)
   } else {
     hint = hintText
-    setTypeface(typeface, Typeface.ITALIC)
+    if (text.isEmpty()) setTypeface(typeface, Typeface.ITALIC)
+    hideSoftKeyboard(v, context)
+  }
+
+  override fun onKeyPreIme(key_code: Int, event: KeyEvent): Boolean {
+    if (event.getKeyCode() == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_UP)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/customview/interaction/TextInputInteractionView.kt]: Ditto as above.

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt [hunk 4]
```diff
@@ -315,6 +315,7 @@ class StateFragmentPresenter @Inject constructor(
 
     viewModel.itemList.clear()
     viewModel.itemList += pendingItemList
+    binding.stateRecyclerView.smoothScrollToPosition(0)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: Check PR #408 for this and merge that in this PR.

File: app/src/main/java/org/oppia/app/utility/KeyBoardHelpher.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package org.oppia.app.utility
+
+import android.content.Context
+import android.view.View
+import android.view.inputmethod.InputMethodManager
+
+class KeyBoardHelpher {
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/utility/KeyBoardHelpher.kt]: typo in this name.

Also rename this to `KeyboardHelper`.

Also, add Kdocs for this class.

File: gradle/wrapper/gradle-wrapper.properties [hunk 6]
```diff
@@ -1,6 +1,6 @@
-#Fri Sep 13 18:15:04 IST 2019
+#Mon Nov 18 10:47:27 IST 2019
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.1.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.4.1-all.zip
```
Comment (rt4914) [inline gradle/wrapper/gradle-wrapper.properties:6]: @nikitamarysolomanpvt Are these changes necessary?

**Review by rt4914 — APPROVED** [review]
LGTM, but make sure you integrate changes from #408 in this PR, before merging this to develop.

**Comment by veena14cs** [comment]
Please take latest update from develop and also add screenshot for reference.

**Review by veena14cs — APPROVED** [review]
LGTM.

File: gradle/wrapper/gradle-wrapper.properties (hunk 6 above)
Comment (nikitamarysolomanpvt) [reply gradle/wrapper/gradle-wrapper.properties:6]: Iam facing issue while using this update from develop of gradle version gradle-5.1.1 . We could raise an issue and fix this if anybody facing issue after upgrading to 5.4.1.

File: app/src/main/java/org/oppia/app/utility/KeyBoardHelpher.kt (hunk 5 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/utility/KeyBoardHelpher.kt]: Done

File: app/src/main/java/org/oppia/app/customview/interaction/NumericInputInteractionView.kt (hunk 2 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/customview/interaction/NumericInputInteractionView.kt]: Done

File: app/src/main/java/org/oppia/app/customview/interaction/FractionInputInteractionView.kt (hunk 1 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/customview/interaction/FractionInputInteractionView.kt]: Done

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt (hunk 4 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: Done

**Comment by nikitamarysolomanpvt** [comment]
@mschanteltc as per your reply in mail **Fix #28,#29,#33: Hi-fi input interaction views #330**


> I agree that when the input is in focus, the keyboard would appear. However if the user taps on the screen that is not the input bar or the keyboard, the keyboard should collapse. If the input is still empty, then it would still show placeholder text. Otherwise if the input has characters in it, those characters remain in the bar and the placeholder text is not seen. 

Please find the your suggestions implemented in this PR.

File: gradle/wrapper/gradle-wrapper.properties (hunk 6 above)
Comment (rt4914) [reply gradle/wrapper/gradle-wrapper.properties:6]: Okay, for now we should merge this.

**Review by rt4914 — APPROVED** [review]
LGTM

File: gradle/wrapper/gradle-wrapper.properties (hunk 6 above)
Comment (BenHenning) [reply gradle/wrapper/gradle-wrapper.properties:6]: @nikitamarysolomanpvt can you include the issue you're facing here for posterity? It's helpful if someone comes back to this comment in the future & wonders why we made the gradle version change.

File: app/src/main/java/org/oppia/app/customview/interaction/FractionInputInteractionView.kt [hunk 7]
```diff
@@ -26,11 +31,19 @@ class FractionInputInteractionView @JvmOverloads constructor(
...
     typeface = Typeface.DEFAULT
+    showSoftKeyboard(v, context)
   } else {
     hint = hintText
-    setTypeface(typeface, Typeface.ITALIC)
+    if (text.isEmpty()) setTypeface(typeface, Typeface.ITALIC)
+    hideSoftKeyboard(v, context)
+  }
+
+  override fun onKeyPreIme(key_code: Int, event: KeyEvent): Boolean {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/customview/interaction/FractionInputInteractionView.kt]: Should be ``keyCode`` since we use lowerCamelCase for identifiers. Ditto for other files changed in this PR (I counted 2 more occurrences).

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt [hunk 8]
```diff
@@ -280,6 +280,12 @@ class StateFragmentPresenter @Inject constructor(
     }
 
     val ephemeralState = result.getOrThrow()
+
+    var scrollToTop = false
+    if (currentStateName.isNotEmpty() && currentStateName != ephemeralState.state.name) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: Let's keep currentStateName a lateinit var. You can check if it's initialized, which I think devolves to something like:

```kotlin
val scrollToTop = ::currentStateName.isInitialized && currentStateName != ephemeralState.state.name
```

Also note that this simplifies scrollToTop and allows it to be a val.

File: app/src/main/java/org/oppia/app/utility/KeyboardHelper.kt [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package org.oppia.app.utility
+
+import android.content.Context
+import android.view.View
+import android.view.inputmethod.InputMethodManager
+
+/** KeyboardHelper help to handle the visiblity of softinputkeybord. */
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/utility/KeyboardHelper.kt]: Nit: ``to change the visibility``. This doesn't really "handle" it like one might expect, but it does let calling code affect the visibility.

File: app/src/main/java/org/oppia/app/utility/KeyboardHelper.kt [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+class KeyboardHelper {
+  companion object {
+    fun hideSoftKeyboard(view: View, context: Context) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/utility/KeyboardHelper.kt:15]: Please add a brief KDoc here & for the next method explaining what the method does, and what view/context to supply.

**Review by BenHenning — APPROVED** [review]
Generally LGTM--thanks @nikitamarysolomanpvt! I had some comments that should definitely be addressed, but the solution seems sensible based on my understanding of focus behavior on Android. Also, I tried it out locally and it seems to work quite well.

**Comment by mschanteltc** [comment]
> @mschanteltc as per your reply in mail **Fix #28,#29,#33: Hi-fi input interaction views #330**
> 
> > I agree that when the input is in focus, the keyboard would appear. However if the user taps on the screen that is not the input bar or the keyboard, the keyboard should collapse. If the input is still empty, then it would still show placeholder text. Otherwise if the input has characters in it, those characters remain in the bar and the placeholder text is not seen.
> 
> Please find the your suggestions implemented in this PR.

Thanks Nikita! Is it possible you can attach an image/gif so I can see how this works please?

File: app/src/main/java/org/oppia/app/customview/interaction/FractionInputInteractionView.kt (hunk 7 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/customview/interaction/FractionInputInteractionView.kt]: Done in all three input view classes

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt (hunk 8 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: Done

File: app/src/main/java/org/oppia/app/utility/KeyboardHelper.kt (hunk 9 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/utility/KeyboardHelper.kt]: Done

File: app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt (hunk 8 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt]: Done

File: app/src/main/java/org/oppia/app/utility/KeyboardHelper.kt (hunk 10 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/utility/KeyboardHelper.kt:15]: Done

**Comment by rt4914** [comment]
@nikitamarysolomanpvt This PR looks good to me including Ben's suggestions. Just reply to this [comment](https://github.com/oppia/oppia-android/pull/405#discussion_r347722189) and Chantel's comment.

**Review by rt4914 — APPROVED** [review]
LGTM.

File: gradle/wrapper/gradle-wrapper.properties (hunk 6 above)
Comment (nikitamarysolomanpvt) [reply gradle/wrapper/gradle-wrapper.properties]: Now i'am not facing this issue. If it occurs again will mention. Was using 5.4.1. only before don't know why it's changed.

Comment (nikitamarysolomanpvt) [reply gradle/wrapper/gradle-wrapper.properties:6]: small

**Comment by nikitamarysolomanpvt** [comment]
@mschanteltc 
> > @mschanteltc as per your reply in mail **Fix #28,#29,#33: Hi-fi input interaction views #330**
> > > I agree that when the input is in focus, the keyboard would appear. However if the user taps on the screen that is not the input bar or the keyboard, the keyboard should collapse. If the input is still empty, then it would still show placeholder text. Otherwise if the input has characters in it, those characters remain in the bar and the placeholder text is not seen.
> > 
> > 
> > Please find the your suggestions implemented in this PR.
> 
> Thanks Nikita! Is it possible you can attach an image/gif so I can see how this works please?

gif

**Comment by nikitamarysolomanpvt** [comment]
> @nikitamarysolomanpvt This PR looks good to me including Ben's suggestions. Just reply to this [comment](https://github.com/oppia/oppia-android/pull/405#discussion_r347722189) and Chantel's comment.

Done

**Comment by mschanteltc** [comment]
When the input is empty, can we make the "Submit" button disabled? Also can we make the blue boxes' bottom be the same height as the blue boxes' top padding?

Otherwise the functionality of the text input LGTM!
